Nc news 11

tr-tobi opened the pull request.

Commit:
implements testing for article queries

Commit:
handles response and catches errors for queries

Commit:
handles errors for article queries

Commit:
handles queries table to order the data based on the given query

Commit:
updates description of article endpoint by adding the queries

Commit:
implements testing for comment_count

Commit:
adds to existing query to collect comment_count

Slyose commented on models/model.js:
```diff
@@ -8,22 +8,52 @@ exports.sendUsers = () => {
 exports.selectArticleById = (article_id) => {
   return db
-    .query("SELECT * FROM articles WHERE article_id = $1;", [article_id])
+    .query(
+      `SELECT a.*, COALESCE(CAST(c.comment_count AS INT), 0) AS comment_count
+      FROM articles a
+      LEFT JOIN (SELECT article_id, CAST(COUNT(*) AS INT) AS comment_count
+    FROM comments
+    GROUP BY article_id) 
+    c ON a.article_id = c.article_id
+    WHERE a.article_id = $1;`,
+      [article_id]
+    )
```
Ideally, this change would be on it's own branch. We want to keep our PRs focused on just one task at a time!

Slyose commented on models/model.js:
```diff
@@ -8,22 +8,52 @@ exports.sendUsers = () => {
 
-exports.sendArticles = () => {
+exports.sendArticles = (topic, sort_by = "created_at", order = "desc") => {
+  const acceptedSort = [
+    "title",
+    "topic",
+    "author",
+    "created_at",
+    "article_img_url",
+  ];
+  const acceptedOrder = ["asc", "desc"];
+  if (!acceptedSort.includes(sort_by) || !acceptedOrder.includes(order)) {
+    return Promise.reject({ status: 400, msg: "Bad Request" });
+  }
```
Good protection against SQL injection here!

Slyose commented on __tests__/app.test.js:
```diff
@@ -98,6 +98,81 @@ describe("/articles", () => {
+      .then((res) => {
+        expect(res.body.msg).toBe("Bad Request");
+      });
+  });
```
Great testing so far! Would be good to see a few more cases covered...
- A status 200; valid topic query but one with no articles. Should respond with an empty array of articles (e.g. ?topic=paper)
- A status 404; non-existent topic query (e.g. ?topic=bananas)

Slyose reviewed (commented):
Nice work! A few more test cases and this will be ready to merge.

Commit:
implements two more tests for the topic query

Commit:
handles error responses

Commit:
checks whether the topic is valid

tr-tobi commented:
Hi, I would like to know what i would do for the production topics because I implemented my code to only get a 200 status when the ?topic query is one of the topics in the test-data/topics.js file. 

Would I need to change the topics manually before hosting the application and with the way i implemented  it, would it make sense to import the topics from the topics data file using a util function?

Slyose commented:
> Hi, I would like to know what i would do for the production topics because I implemented my code to only get a 200 status when the ?topic query is one of the topics in the test-data/topics.js file.
> 
> Would I need to change the topics manually before hosting the application and with the way i implemented it, would it make sense to import the topics from the topics data file using a util function?

Hey Tobi! Great question. In the case of the user passing a topic, we should query the db for that given topic in the model and look at the response. If we find that topic, then we can progress with the request, and if not we should go ahead and promise.reject it.

This is the best way to do it because your seed data in a real project would likely end up being stale & out of date since a database changes often

Commit:
removes consolelog

Commit:
makes topic array more dynamic

Slyose commented on models/model.js:
```diff
@@ -8,22 +10,62 @@ exports.sendUsers = () => {
+  if (!acceptedSort.includes(sort_by) || !acceptedOrder.includes(order)) {
+    return Promise.reject({ status: 400, msg: "Bad Request" });
+  }
+  const uniqueTopics = [];
+  testTopics.forEach((item) => {
+    if (!uniqueTopics.includes(item.slug)) {
+      uniqueTopics.push(item.slug);
+    }
+  });
```
This method wouldn't be ideal because you're not querying the actual database - the seed data json file could become stale & out of date.

To make it truly dynamic, if the user passes a topic, do a db query to check if that topic is actually in the DB at the time of query ( rejecting the promise if it isnt)

image if at some point another topic was added to the db from elsewhere, you wouldn't be able to query that topic because it's not in the seed data

Slyose reviewed (commented):
Unfortunately the way you handed uniqueTopics isn't ideal here, please take a look at the feedback and refactor. Almost there with this ticket, just needs an actual db.query to check for topics instead of a util function

Commit:
queries for list of topics to make function more dynamic

Augs0 commented on __tests__/app.test.js:
```diff
@@ -98,6 +98,97 @@ describe("/articles", () => {
+            "https://images.pexels.com/photos/158651/news-newsletter-newspaper-information-158651.jpeg?w=700&h=700",
+        });
+        expect(res.body.articles[0].topic).toBe("cats");
+        expect(res.body.articles).toBeSorted("created_at");
```
We should also assert that it's descending as that's the default order

Augs0 commented on __tests__/app.test.js:
```diff
@@ -98,6 +98,97 @@ describe("/articles", () => {
+        expect(res.body.articles).toBeSorted("created_at");
+      });
+  });
+  test("GET:200 sends an array of the article objects to the end point by topic value  specified in the query", () => {
```
How is this different to the test above?

Augs0 commented on __tests__/app.test.js:
```diff
@@ -98,6 +98,97 @@ describe("/articles", () => {
+        });
+      });
+  });
+  test("GET:200 sends an array of the article objects to the end point in descending order", () => {
```
This also seems to be a repeat apart from the descending assertion which could have been included above as you're still testing the default value

Augs0 reviewed (approved):
You seem to have implemented the feedback Modou gave, well done! I've left a few comments where you seem to be repeating yourself a bit , but otherwise you are ok to merge

Commit:
removes repeat assertions
